src/stdlib/ListUtils.scillib

```diff
@@ -245,47 +278,29 @@ let list_sort : forall 'A. ('A -> 'A -> Bool) -> List 'A -> List 'A =
 (* Return None iff none of the elements in "l" satisfy "f". *)
 let list_find : forall 'A. ('A -> Bool) -> List 'A -> Option 'A =
   tfun 'A =>
-  fun (f : 'A -> Bool) =>
-  fun (l : List 'A) =>
-    let folder = @list_foldl 'A (Option 'A) in
-    let init = None {'A} in
-    let iter =
-      fun (z : Option 'A) =>
-      fun (h : 'A) =>
-        match z with
-        | Some a =>
-          z
-        | None =>
-          let r = f h in
-          match r with
-          | True =>
-            Some {'A} h
-          | False =>
-            None {'A}
-          end
-        end
-     in
-       folder iter init l
+  fun (p : 'A -> Bool) => fun (ls : List 'A) =>
+  let none = None {'A} in
+  (* continue fold on None, exit fold when Some compare st. p(compare) *)
+  let predicate_step = fun (ignore : Option 'A) => fun (compare : 'A) =>
+  fun (recurse: Option 'A -> Option 'A) =>
+    let satisfies_p = p compare in
+    match satisfies_p with
+    | True => Some {'A} compare
+    | False => recurse none
+    end in
+  let foldk = @list_foldk 'A (Option 'A) in
+  foldk predicate_step none ls
 
 (* Return True iff at least one element of list "l" satisfy predicate "f". *)
 let list_exists : forall 'A. ('A -> Bool) -> List 'A -> Bool =
   tfun 'A =>
-  fun (f : 'A -> Bool) =>
-  fun (l : List 'A) =>
-    let folder = @list_foldl 'A Bool in
-    let init = False in
-    let iter =
-      fun (z : Bool) =>
-      fun (h : 'A) =>
-        let res = f h in
-        match res with
-        | True =>
-          True
-        | False =>
-          z
-        end
-    in
-      folder iter init l
+  fun (p : 'A -> Bool) => fun (ls: List 'A) =>
```

Comment:
would it consume less gas to implement this from scratch using `list_foldk` rather than using `list_find` ?